src/Microsoft.ML.Core/Data/ColumnType.cs

```diff
@@ -57,42 +53,49 @@ internal ColumnType(Type rawType, DataKind rawKind)
             rawType.TryGetDataKind(out tmp);
             Contracts.Assert(tmp == rawKind);
 #endif
-            _rawType = rawType;
-            _rawKind = rawKind;
+            RawType = rawType;
+            RawKind = rawKind;
         }
 
         /// <summary>
-        /// The raw System.Type for this ColumnType. Note that this is the raw representation type
-        /// and NOT the complete information content of the ColumnType. Code should not assume that
-        /// a RawType uniquely identifiers a ColumnType.
+        /// The raw <see cref="Type"/> for this <see cref="ColumnType"/>. Note that this is the raw representation type
+        /// and not the complete information content of the <see cref="ColumnType"/>. Code should not assume that
+        /// a <see cref="RawType"/> uniquely identifiers a <see cref="ColumnType"/>. For example, most practical instances of
+        /// <see cref="KeyType"/> and <see cref="NumberType.U4"/> will have a <see cref="RawType"/> of <see cref="uint"/>,
+        /// but both are very different in the types of information conveyed in that number.
         /// </summary>
-        public Type RawType { get { return _rawType; } }
+        public Type RawType { get; }
 
         /// <summary>
         /// The DataKind corresponding to RawType, if there is one (zero otherwise). It is equivalent
         /// to the result produced by DataKindExtensions.TryGetDataKind(RawType, out kind).
         /// </summary>
-        public DataKind RawKind { get { return _rawKind; } }
+        [BestFriend]
```

Comment:
I think it would be better if we were able to design ColumnType in such a way that it didn't need `[BestFriend] internal` on any of its members.  Imagine when we try to extract IDataView from ML.NET. We will need to redesign this type again in order to get rid of this internal usage of ColumnType to the rest of ML.NET.

It would be better to design it up front in such a way that would work anywhere.

For example, the comment below: ` External code should use <c>is <see cref="PrimitiveType"/></c>.` - why isn't that good enough for everyone?

Reply:
> It would be better to design it up front in such a way that would work anywhere.

Hi @eerhardt , it does not *need* any of these internal conveniences, since everything you can get through the internal properties can be fetched through the public properties. So let's first distinguish whether we *could* use the public surface internally (yes, in case not clear) from the question of whether we should do that work right now.

I think it's pretty clear we could, so you are I guess insisting we do that right now. Perhaps you're not aware of the amount of effort entailed by this request, so I will enumerate it.

If I right now just remove all these things in my change where I added this `BestFriend` attribute in this file, I get 1549 errors. Each one of those requires, based on my experience in doing the tests yesterday when I was preparing this PR, about one minute to resolve each as I think about how to restructure the code so that it is equivalent -- the changes are not mechanical. They are *easy*, but not *mechanical* -- great first issues, by the way. 26 hours of work, let's say. Factor in breaks, meetings, and other work, I'd be surprised if I could actually reasonably get that done in a week.

I was actually kind of hoping to be done with this *today* so that I could move on to hiding many other public surfaces in time for the API review. Now then, that's not going to happen for various reasons (since @Zruty0 is the only other one that has signed off so far), which is a pity, but whatever.

It may be that having *someone* do this work eventually would be a good idea for the reasons you have outlined, but I'd say in this precise moment insisting *I* do it when there are many other things that almost certainly require my unique attention would not be wise. What I proposed in #1533, and which I still believe is the wiser course of action, was that I *not* do that, and instead make the public surface what we want it to be.